### PR TITLE
switch to dynamic language refmt

### PR DESCRIPTION
**Pull request by rickyvetter**
This allows for flexible conversion between ML/RE and to itself. It will open up the ability to do RE/ML=>JS conversation in a future iteration. I think I'd also like to get line width as an option for the output.

screen shot 2017-04-25 at 11 02 15 pm

This diff looks way bigger than it is because of refmt changes :(

### CODE REVIEW

File: src/extension/popup/PopupCommon.re [hunk 1]
```diff
@@ -31,24 +31,64 @@ let makeContentHash text => "#" ^ Util.btoa text;
...
+    let hash = makeContentHash input;
+    let link = generateShareableLink hash;
+    let inLang = e |> ReactEventRe.Synthetic.target |> LocalDom.Element.value;
+    refmt input ::inLang (fun outText inLang outLang => render input outText inLang outLang link)
+  }
+  and outLangChanged input e => {
+    let hash = makeContentHash input;
+    let link = generateShareableLink hash;
+    let outLang = e |> ReactEventRe.Synthetic.target |> LocalDom.Element.value;
+    refmt input ::outLang (fun outText inLang outLang => render input outText inLang outLang link)
```
Comment (glennsl) [inline src/extension/popup/PopupCommon.re]: There's not a whole lot of difference between `inLangChanged` and `outLangChanged`. A couple of helpers, one that wraps the refmt call and computes `link` and `hash`, and one that retrieves the value would be good perhaps.

File: src/extension/popup/PopupCommon.re [hunk 2]
```diff
@@ -31,24 +31,64 @@ let makeContentHash text => "#" ^ Util.btoa text;
...
+  }
+  and inputChanged ::inLang="unknown" ::outLang="unknown" input => {
     let hash = makeContentHash input;
     let link = generateShareableLink hash;
```
Comment (glennsl) [inline src/extension/popup/PopupCommon.re]: Here too is some duplication

File: src/extension/popup/PopupWindow.re [hunk 3]
```diff
@@ -33,20 +85,52 @@ module PopupWindow = {
...
+    props.onInputChanged
+      inLang::(stringOfLanguageSetting inLanguage)
+      outLang::(stringOfLanguageSetting state.outLanguage)
+      props.inText;
+    Some {...state, inLanguage}
+  };
+  let handleOutLangChange {props, state} e => {
+    let outLanguage =
+      e |> ReactEventRe.Synthetic.target |> LocalDom.Element.value |> languageSettingOfString;
+    props.onInputChanged
```
Comment (glennsl) [inline src/extension/popup/PopupWindow.re]: What's `inLangChanged` and `outLangChanged` for if they're not used here? Also, lots of duplication here as well, seems ripe for some helpers.

File: src/extension/popup/PopupWindow.re [hunk 4]
```diff
@@ -33,20 +85,52 @@ module PopupWindow = {
...
+      input;
+  let render {props, state, updater, handler} => {
+    Js.log props;
     <div style=PopupStyles.popup>
       <div style=PopupStyles.popupColumn>
-        <h1 style=PopupStyles.popupContext> <ColumnTitle name="In" lang=props.inLang /> </h1>
+        <h1 style=PopupStyles.popupContext>
+          <ColumnTitle
+            lang=props.inLang
+            select=(select "in" (updater handleInLangChange) state.inLanguage props.inLang)
```
Comment (glennsl) [inline src/extension/popup/PopupWindow.re]: It's not very clear here that `select` is a higher order component. I think I'd prefer `select` to be a barebones JSX component wrapped in a lambda.

File: src/refmt/Refmt.re [hunk 5]
```diff
@@ -71,50 +100,107 @@ let genErrors errors => {
...
-let refmtJS jsString =>
-  switch (refmt (Js.to_string jsString)) {
-  | MLtoRE s => Js.array [|Js.string "MLtoRE", Js.string s|]
-  | REtoML s => Js.array [|Js.string "REtoML", Js.string s|]
-  | MLItoREI s => Js.array [|Js.string "MLItoREI", Js.string s|]
-  | REItoMLI s => Js.array [|Js.string "REItoMLI", Js.string s|]
-  | Failure s => Js.array [|Js.string "Failure", Js.string s|]
-  };
+let refmtJS jsString inLang inType outLang => {
+  let (trueIn, trueOut, s) =
```
Comment (glennsl) [inline src/refmt/Refmt.re]: Could `actualIn` and `actualOut` be better names here?

Comment (rickyvetter) [inline src/refmt/Refmt.re]: 👍

File: src/extension/popup/PopupWindow.re (hunk 4 above)
Comment (rickyvetter) [inline src/extension/popup/PopupWindow.re]: 👍

File: src/extension/popup/PopupWindow.re (hunk 3 above)
Comment (rickyvetter) [inline src/extension/popup/PopupWindow.re]: Ahh yeah the ones in the early file can be deleted I believe!

**Comment by rickyvetter** [comment]
Thanks for the review! I will delete/clean up some of this code and look into minimizing the variant/string conversions happening.

**Pull request merged**